### PR TITLE
Cleaning up the FCS code

### PR DESCRIPTION
**Pull request by VKing6**
Another thing that started out as a relatively simple fix has spiralled out of control as I make way too many branches that inevitably start intermingling with rewrites and cleanups from here to Frankfurt.

### CODE REVIEW

File: addons/fcs/functions/fnc_getRange.sqf [hunk 1]
```diff
@@ -12,7 +14,26 @@
...
 
+params [["_accuracy",1], ["_maxDistance",5000], ["_minDistance",0]];
+
 disableSerialization;
 private _dlgRangefinder = uiNamespace getVariable ["ACE_dlgRangefinder", displayNull];
 
-parseNumber ctrlText (_dlgRangefinder displayCtrl 151);
+private _distance = parseNumber ctrlText (_dlgRangefinder displayCtrl 151);
+
+if (isNil "_distance" || _distance == 0) then {
```
Comment (commy2) [inline addons/fcs/functions/fnc_getRange.sqf]: _distance is quaranteed to be defined.

Comment (VKing6) [inline addons/fcs/functions/fnc_getRange.sqf]: Even if the vehicle can't get range via the rangefinder?

Comment (jonpas) [inline addons/fcs/functions/fnc_getRange.sqf]: This would need lazy evaluation as well, or it will error on OR.

**Comment by commy2** [comment]
Worst case would be:
`parseNumber ctrlText (<displayNull> displayCtrl 151);`
->
`parseNumber ctrlText <controlNull>`
->
`parseNumber <"">`
->
`0`

**Comment by VKing6** [comment]
Fair enough.

**Comment by VKing6** [comment]
Okay I've gone a bit overboard again, but at least the following things have happened:
* InitSpeed is now properly determined, until BIS fixes something.
* Config values for minimum, maximum ranges and measurement accuracy are now adhered to.
* Laser range display is now less spastic

**Comment by esteldunedain** [comment]
:+1: Looks good as far as I can see.

There are quite a few oportunities for caching config reads but they are only used when pressing or releasing the range key, so it's probably not that critical.

File: addons/fcs/functions/fnc_calculateSolution.sqf [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    private _magazine = _x;
+    private _ammo = getText (configFile >> "CfgMagazines" >> _magazine >> "ammo");
+
+    if !(getText (configFile >> "CfgAmmo" >> _ammo >> "simulation") == "shotMissile") then {
+        private _maxElev     = getNumber (_turretConfig >> "maxElev");
+        private _initSpeed   = getNumber (configFile >> "CfgMagazines" >> _magazine >> "initSpeed");
+        private _airFriction = getNumber (configFile >> "CfgAmmo" >> _ammo >> "airFriction");
+
+        {
+            private ["_weapon", "_muzzles", "_weaponMagazines", "_muzzleMagazines"];
```
Comment (Glowbal) [inline addons/fcs/functions/fnc_calculateSolution.sqf]: Could probably use the new private syntax for these?

File: addons/fcs/functions/fnc_calculateSolution.sqf [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            _weapon = _x;
+            _muzzles = getArray (configFile >> "CfgWeapons" >> _weapon >> "muzzles");
+            _weaponMagazines = getArray (configFile >> "CfgWeapons" >> _weapon >> "magazines");
+
+            {
+                if (_x != "this") then {
+                    _muzzleMagazines = getArray (configFile >> "CfgWeapons" >> _weapon >> _x >> "magazines");
+                    _weaponMagazines append _muzzleMagazines;
```
Comment (Glowbal) [inline addons/fcs/functions/fnc_calculateSolution.sqf]: Thise can be done without the `_muzzleMagazines` variable assignment

File: addons/fcs/functions/fnc_getRange.sqf [hunk 4]
```diff
@@ -12,7 +15,35 @@
...
 
-parseNumber ctrlText (_dlgRangefinder displayCtrl 151);
+private _distance = parseNumber ctrlText (_dlgRangefinder displayCtrl 151);
+
+if (_distance == 0) then {
+    _distance = _this call EFUNC(common,getTargetDistance);
+} else {
+    if (_distance > _maxDistance) then {
+        _distance = _maxDistance;
+    };
```
Comment (Glowbal) [inline addons/fcs/functions/fnc_getRange.sqf]: `_distance = _distance max _madDistance` ?

File: addons/fcs/functions/fnc_getRange.sqf [hunk 5]
```diff
@@ -12,7 +15,35 @@
...
+
+    if (_distance < _minDistance) then {
+        _distance = _minDistance;
+    };
+
+    _accuracy = if(_accuracy < 1) then {1} else {_accuracy};
```
Comment (Glowbal) [inline addons/fcs/functions/fnc_getRange.sqf]: Same as above

File: addons/fcs/functions/fnc_keyUp.sqf [hunk 6]
```diff
@@ -109,61 +109,11 @@ if (_viewDiff != 0) then {
...
-    };
-    false
-} count (_vehicle magazinesTurret _turret);
-
-[_vehicle, format ["%1_%2", QGVAR(Distance),  _turret],     _distance] call EFUNC(common,setVariablePublic);
-[_vehicle, format ["%1_%2", QGVAR(Magazines), _turret], _FCSMagazines] call EFUNC(common,setVariablePublic);
-[_vehicle, format ["%1_%2", QGVAR(Elevation), _turret], _FCSElevation] call EFUNC(common,setVariablePublic);
 [_vehicle, format ["%1_%2", QGVAR(Azimuth),   _turret],   _FCSAzimuth] call EFUNC(common,setVariablePublic);
 
+// CALCULATE SOLUTION
```
Comment (Glowbal) [inline addons/fcs/functions/fnc_keyUp.sqf]: Could probably get rid of this comment. Its the name of the method after all

Comment (VKing6) [inline addons/fcs/functions/fnc_keyUp.sqf]: I kinda like having it there as it marks the end of the azimuth section.

**Pull request merged**